host/sgx/switchless.c

```diff
@@ -29,11 +29,17 @@ static void* _switchless_ocall_worker(void* arg)
         volatile oe_call_host_function_args_t* local_call_arg = NULL;
         if ((local_call_arg = context->call_arg) != NULL)
         {
-            context->call_arg = NULL;
```

Comment:
Help me understand this: the performance is improved because we avoid putting switchless calls on a thread that could stall on the current switchless call waiting to be finished, right? On the other hand, I image this would increase the chance that a switchless call has to fall back to the regular call. 

In this test case, there is only one work thread. So your change would guarantee the switchless call must fall back to regular call if the work thread is handling another switchless call. Are you suggesting the fall back actually improved performance?

Reply:
> 
> 
> Help me understand this: the performance is improved because we avoid putting switchless calls on a thread that could stall on the current switchless call waiting to be finished, right?
>

Yes. The non-optimality of the current scheduling is that no matter the number of host workers, enclave threads will stall since calls will be scheduled on a worker threads before it has finished handling the previous call.

**There is no way for the user to over come this and the user cannot achieve maximum throughput.**

>
>On the other hand, I image this would increase the chance that a switchless call has to fall back to the regular call.
> 

Since our switchless ocalls are blocking, **fallback will never happen when the number of host worker threads >= number of enclave threads making ecalls**. In this case, there will always be a host worker available.

With the new scheduling, in the case where number of worker threads < number of enclave threads making ecalls, fallback can happen. However, fallback still allows the system to make progress whereas stalling does not.

Note that if number of enclave threads > 2* worker threads, then fallback will happen in the current implementation as well.

>
>In this test case, there is only one work thread. 
>
This test allows passing in different number of host and enclave threads. 

>
> Are you suggesting the fall back actually improved performance?
>
No. Not stalling improved performance.

Reply:
That's my intuition: this PR will make fallback happens more often. Imaging a switchless call comes in, and find all work threads are occupied, it either 1) wait a while on the first thread it poked, or 2) fall back to the regular call. If I remember correctly, the regular call is 5-10 times slower than switchless call. So I'd not automatically conclude 2) is better than 1). 

I'd be more convinced with a print out of the speed up under difference scenarios, especially when enclave thread# > worker thread#.

Reply:
BTW, I think the problem with the non-optimal current scheduler is that we always check with the first work thread, if it's busy, then go on to the 2nd one, and so on... If we "load balance" somehow, for example, this time starts with #1 thread and continue forward, next time starts with the last thread and go backward, the performance should be better.

Reply:
@jxyang 
For my understanding:
- Is the current scheduling strategy of stalling instead of using another available worker explicitly chosen to minimize fallback? 
- Fallback will happen only if number of worker threads is less than number of enclave threads. Do we anticipate that for using the switchless calls feature, the enclave author will use less worker threads than enclave threads?

>
>  If we "load balance" somehow, for example, this time starts with #1 thread and continue forward, next time starts with the last thread and go backward, the performance should be better.
>

I am not sure I understand your proposal. Can you suggest a complete algorithm?

Reply:
Consider a simple scenario with 2 worker threads and 2 enclave threads

Here is what can happen with current scheduling:
1. enc-thread1 schedules call in worker1
2. enc-thread2 chooses worker1 and stalls waiting for worker1 to complete
3. worker1 completes and starts to process enc-thread2 call
4. enc-thread1 chooses worker 1 and stalls waiting for worker 1 to complete
5. worker1 completes and starts process enc-thread1 call
6. enc-thread2 chooses worker 1 and stalls waiting for worker 1 to complete
...

worker-2 is never used.

Is there are better way to leverage worker-2 than what is being done in this PR?